streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -602,6 +608,47 @@ private void removeUnusedTaskFromStateUpdater(final TaskId taskId) {
         tasks.addPendingTaskToCloseClean(taskId);
     }
 
+    private void addToTasksToClose(final Map<TaskId, CompletableFuture<StateUpdater.RemovedTaskResult>> futures,
+                                   final Set<Task> tasksToCloseCleanFromStateUpdater,
+                                   final Set<Task> tasksToCloseDirtyFromStateUpdater) {
+        iterateAndActOnFuture(futures, removedTaskResult -> {
+            final Task task = removedTaskResult.task();
+            final Optional<RuntimeException> exception = removedTaskResult.exception();
+            if (exception.isPresent()) {
+                tasksToCloseDirtyFromStateUpdater.add(task);
+            } else {
+                tasksToCloseCleanFromStateUpdater.add(task);
+            }
+        });
+    }
+
+    private void iterateAndActOnFuture(final Map<TaskId, CompletableFuture<StateUpdater.RemovedTaskResult>> futures,
+                                       final java.util.function.Consumer<StateUpdater.RemovedTaskResult> action) {
+        for (final Map.Entry<TaskId, CompletableFuture<StateUpdater.RemovedTaskResult>> entry : futures.entrySet()) {
+            final TaskId taskId = entry.getKey();
+            final CompletableFuture<StateUpdater.RemovedTaskResult> future = entry.getValue();
+            try {
+                final StateUpdater.RemovedTaskResult removedTaskResult = waitForFuture(taskId, future);
+                action.accept(removedTaskResult);
+            } catch (final ExecutionException executionException) {
+                log.warn("An exception happened when removing task {} from the state updater. The exception will be handled later: ",
+                    taskId, executionException);
+            } catch (final InterruptedException ignored) { }
```

Comment:
We are quite inconsistent on how we treat `InterruptedException`. In some places we ignore them because they should not happen and in others we treat them as fatal and throw an `IllegalStateException` because they should not happen [1].

[1] https://github.com/apache/kafka/blob/b36cf4ef977fb14bc57683630a9f3f3680705550/streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java#L597

Reply:
I am actually in favor of treating them as fatal and throw an `IllegalStateException` to make it more explicit that interruption of a stream thread should not happen. WDYT?

Reply:
Sounds good!